README.md
# flask-sqlachemy-api

generate API Rest from Model Class


## Installation


    pip install flask-sqlachemy-api
        
Or

    git clone https://github.com/fraoustin/flask-sqlachemy-api.git
    cd flask-sqlachemy-api
    python setup.py install

You can load test by

    flake8 --ignore E501,E226,E128,F401
    python -m unittest discover -s tests


## Usage

    import os, logging
    from flask import Flask
    from flask_sqlalchemy import SQLAlchemy
    from flask_sqlalchemy_api import ApiRest, error_api

    app = Flask(__name__)

    # db SQLAlchemy
    database_file = "sqlite:///{}".format(os.path.join('.', "test.db"))
    app.config["SQLALCHEMY_DATABASE_URI"] = database_file
    app.config["SQLALCHEMY_TRACK_MODIFICATIONS"] = False
    db = SQLAlchemy() 

    class Todo(db.Model):
        __tablename__ = 'todo'

        id = db.Column(db.Integer, primary_key=True, autoincrement=True)
        title = db.Column(db.String, nullable=False)
        description = db.Column(db.String, nullable=True)
        status = db.Column(db.String, nullable=True)


    apiManager = ApiRest(db)
    for method in ['ALL', 'POST', 'GET', 'DELETE', 'PUT', 'PATCH']:
        apiManager.add_api(Todo, method)
    app.register_blueprint(apiManager)


    if __name__ == "__main__":
        db.init_app(app)
        with app.app_context():
            db.create_all()
        app.logger.setLevel(logging.DEBUG)
        app.run(host='0.0.0.0', port=5000, debug=True)

This code generate api url:

- GET http://127.0.0.1:5000/api/v1/todos
- POST http://127.0.0.1:5000/api/v1/todo
- GET http://127.0.0.1:5000/api/v1/todo/<id>
- DELETE http://127.0.0.1:5000/api/v1/todo/<id>
- PUT http://127.0.0.1:5000/api/v1/todo/<id>
- PATCH http://127.0.0.1:5000/api/v1/todo/<id>


flask-sqlachemy-api generate 6 Apis from Model Class

- ALL: it's a get request for list of item with url http://domain/api_path/items. You can add parameter filter, orderby, offset, limit.

    http://127.0.0.1:5000/api/v1/todos?orderby=title%20desc  method GET
    http://127.0.0.1:5000/api/v1/todos?orderby=title%20desc&filter=id%3D  method GET
    http://127.0.0.1:5000/api/v1/todos?offset=50&limit=50  method GET

- GET: it's get request for  a specific item with url http://domain/api_path/item/<id>

    http://127.0.0.1:5000/api/v1/todo/1 method GET

- DELETE: it's delete request for a specific item url http://domain/api_path/item/<id>

    http://127.0.0.1:5000/api/v1/todo/1  method DELETE

- POST: it's post request for add a item with url http://domain/api_path/item. You add data on your request

    http://127.0.0.1:5000/api/v1/todo  method POST

- PUT: it's put request for modify a specific item with url http://domain/api_path/item/<id>. You add data on your request

    http://127.0.0.1:5000/api/v1/todo/1  method PUT

- PATCH: it's patch request for modify a specific item on specific column with url http://domain/api_path/item/<id>. You add data on your request

    http://127.0.0.1:5000/api/v1/todo/1  method PATCH

You can 

- specific url (default /api/v1) on ApiRest
- add decorator (login, ...) in sample/sample1.py
- specific endpoint
- specific serialize: transform item to json


## TODO

- voir pour faire des apis lié au fk ex /api/person/(int: id)/computers (https://flask-restless.readthedocs.io/en/stable/requestformat.html)

CHANGES.md
# Feature

- generate documentation
- TODO

# V. 0.9.2

- change README


# V. 0.9.1

- correction setup

# V. 0.9.0

- init repos
- add api 'ALL', 'POST', 'GET', 'DELETE', 'PUT', 'PATCH'
- manage decorator, serialize, error

tests/test01_basic.py
import unittest
from flask import Flask, request, json

from flask_sqlalchemy import SQLAlchemy
from flask_sqlalchemy_api import ApiRest, error_api


class BasicTest(unittest.TestCase):
    """
        Class for Basic Unitaire Test for flask_sqlalchemy_api
    """
    def setUp(self):
        self.app = Flask(__name__)
        self.app.testing = True
        self.app.config["SQLALCHEMY_DATABASE_URI"] = "sqlite://"
        db = SQLAlchemy()

        class Todo(db.Model):
            __tablename__ = 'todo'
            id = db.Column(db.Integer, primary_key=True, autoincrement=True)
            title = db.Column(db.String, nullable=False)
            description = db.Column(db.String, nullable=True)
            status = db.Column(db.String, nullable=True)

        apiManager = ApiRest(db)
        for method in ['ALL', 'POST', 'GET', 'DELETE', 'PUT', 'PATCH']:
            apiManager.add_api(Todo, method)
        self.app.register_blueprint(apiManager)

        db.init_app(self.app)
        with self.app.app_context():
            db.create_all()

    def test_get_null(self):
        with self.app.test_client() as c:
            rv = c.get('/api/v1/todos')
            self.assertEqual(rv.status_code, 200)
            self.assertEqual(json.loads(rv.data), [])

    def test_post(self):
        with self.app.test_client() as c:
            todo = {'title': 'test', 'description': 'test description', 'status': 'test'}
            rv = c.post('/api/v1/todo', data=todo)
            self.assertEqual(rv.status_code, 201)
            result = json.loads(rv.data)
            for key in todo:
                self.assertEqual(result[key], todo[key])
            self.assertEqual(result["id"], 1)

    def test_post_force_increment(self):
        with self.app.test_client() as c:
            todo = {'title': 'test', 'description': 'test description', 'status': 'test', 'id': 2}
            rv = c.post('/api/v1/todo', data=todo)
            self.assertEqual(rv.status_code, 201)
            result = json.loads(rv.data)
            todo['id'] = 1
            self.assertEqual(result, todo)

    def test_get_item(self):
        with self.app.test_client() as c:
            todo = {'title': 'test', 'description': 'test description', 'status': 'test'}
            rv = c.post('/api/v1/todo', data=todo)
            rv = c.get('/api/v1/todo/1')
            self.assertEqual(rv.status_code, 200)
            result = json.loads(rv.data)
            self.assertEqual(result["id"], 1)
            result.pop("id")
            self.assertEqual(result, todo)

    def test_get_items(self):
        with self.app.test_client() as c:
            todoOne = {'title': 'test1', 'description': 'test description1', 'status': 'test'}
            todoTwo = {'title': 'test2', 'description': 'test description2', 'status': 'test'}
            rv = c.post('/api/v1/todo', data=todoOne)
            rv = c.post('/api/v1/todo', data=todoTwo)
            rv = c.get('/api/v1/todos')
            self.assertEqual(rv.status_code, 200)
            result = json.loads(rv.data)
            self.assertEqual(len(result), 2)
            todoOne["id"] = 1
            self.assertEqual(result[0], todoOne)
            todoTwo["id"] = 2
            self.assertEqual(result[1], todoTwo)

    def test_get_item_not_found(self):
        with self.app.test_client() as c:
            rv = c.get('/api/v1/todo/1')
            self.assertEqual(rv.status_code, 400)

    def test_del_items(self):
        with self.app.test_client() as c:
            todoOne = {'title': 'test1', 'description': 'test description1', 'status': 'test'}
            todoTwo = {'title': 'test2', 'description': 'test description2', 'status': 'test'}
            rv = c.post('/api/v1/todo', data=todoOne)
            rv = c.post('/api/v1/todo', data=todoTwo)
            rv = c.get('/api/v1/todos')
            self.assertEqual(rv.status_code, 200)
            result = json.loads(rv.data)
            self.assertEqual(len(result), 2)
            rv = c.delete('/api/v1/todo/1')
            self.assertEqual(rv.status_code, 200)
            rv = c.get('/api/v1/todos')
            result = json.loads(rv.data)
            self.assertEqual(len(result), 1)
            todoTwo["id"] = 2
            self.assertEqual(result[0], todoTwo)

    def test_delete_item_not_found(self):
        with self.app.test_client() as c:
            rv = c.delete('/api/v1/todo/1')
            self.assertEqual(rv.status_code, 400)

    def test_put_items(self):
        with self.app.test_client() as c:
            todo = {'title': 'test1', 'description': 'test description1', 'status': 'test'}
            rv = c.post('/api/v1/todo', data=todo)
            todo['title'] = 'test1 change'
            todo['description'] = 'test description1 change'
            todo['status'] = 'test change'
            rv = c.put('/api/v1/todo/1', data=todo)
            rv = c.get('/api/v1/todo/1')
            result = json.loads(rv.data)
            todo["id"] = 1
            self.assertEqual(result, todo)

    def test_put_item_not_found(self):
        with self.app.test_client() as c:
            todo = {'title': 'test1', 'description': 'test description1', 'status': 'test'}
            rv = c.put('/api/v1/todo/1', data=todo)
            self.assertEqual(rv.status_code, 400)

    def test_put_items_force_id(self):
        with self.app.test_client() as c:
            todo = {'title': 'test1', 'description': 'test description1', 'status': 'test'}
            rv = c.post('/api/v1/todo', data=todo)
            self.assertEqual(rv.status_code, 201)
            todo['title'] = 'test1 change'
            todo['description'] = 'test description1 change'
            todo['status'] = 'test change'
            todo["id"] = 2
            rv = c.put('/api/v1/todo/1', data=todo)
            rv = c.get('/api/v1/todo/1')
            result = json.loads(rv.data)
            todo["id"] = 1
            self.assertEqual(result, todo)

    def test_patch_items(self):
        with self.app.test_client() as c:
            todo = {'title': 'test1', 'description': 'test description1', 'status': 'test'}
            rv = c.post('/api/v1/todo', data=todo)
            todo['title'] = 'test1 change'
            rv = c.patch('/api/v1/todo/1', data={'title': todo['title']})
            rv = c.get('/api/v1/todo/1')
            result = json.loads(rv.data)
            todo["id"] = 1
            self.assertEqual(result, todo)

    def test_patch_items_check_id(self):
        with self.app.test_client() as c:
            todo = {'title': 'test1', 'description': 'test description1', 'status': 'test'}
            rv = c.post('/api/v1/todo', data=todo)
            todo['title'] = 'test1 change'
            rv = c.patch('/api/v1/todo/1', data={'title': todo['title'], 'id': 2})
            rv = c.get('/api/v1/todo/1')
            result = json.loads(rv.data)
            todo["id"] = 1
            self.assertEqual(result, todo)

    def test_patch_item_not_found(self):
        with self.app.test_client() as c:
            todo = {'title': 'test1'}
            rv = c.patch('/api/v1/todo/1', data=todo)
            self.assertEqual(rv.status_code, 400)

    def test_get_offset_limit(self):
        with self.app.test_client() as c:
            todo = {'title': 'test', 'description': 'test description', 'status': 'test'}
            for i in range(0, 1000):
                rv = c.post('/api/v1/todo', data=todo)
            rv = c.get('/api/v1/todos')
            self.assertEqual(rv.status_code, 200)
            result = json.loads(rv.data)
            self.assertEqual(len(result), 999)
            self.assertEqual(result[0]['id'], 1)
            self.assertEqual(result[998]['id'], 999)
            rv = c.get('/api/v1/todos?offset=50&limit=50')
            self.assertEqual(rv.status_code, 200)
            result = json.loads(rv.data)
            self.assertEqual(len(result), 50)
            self.assertEqual(result[0]['id'], 51)
            self.assertEqual(result[49]['id'], 100)

    def test_get_order_by(self):
        with self.app.test_client() as c:
            todoOne = {'title': 'before', 'description': 'test description1', 'status': 'test'}
            todoTwo = {'title': 'after', 'description': 'test description2', 'status': 'test'}
            rv = c.post('/api/v1/todo', data=todoOne)
            rv = c.post('/api/v1/todo', data=todoTwo)
            rv = c.get('/api/v1/todos?orderby=id')
            self.assertEqual(rv.status_code, 200)
            result = json.loads(rv.data)
            self.assertEqual(len(result), 2)
            self.assertEqual(result[0]['id'], 1)
            self.assertEqual(result[1]['id'], 2)
            rv = c.get('/api/v1/todos?orderby=title')
            self.assertEqual(rv.status_code, 200)
            result = json.loads(rv.data)
            self.assertEqual(len(result), 2)
            self.assertEqual(result[0]['id'], 2)
            self.assertEqual(result[1]['id'], 1)
            rv = c.get('/api/v1/todos?orderby=title%20desc')
            self.assertEqual(rv.status_code, 200)
            result = json.loads(rv.data)
            self.assertEqual(len(result), 2)
            self.assertEqual(result[0]['id'], 1)
            self.assertEqual(result[1]['id'], 2)
            rv = c.post('/api/v1/todo', data=todoOne)
            rv = c.get('/api/v1/todos?orderby=title%20desc%2Cid%20desc')
            self.assertEqual(rv.status_code, 200)
            result = json.loads(rv.data)
            self.assertEqual(len(result), 3)
            self.assertEqual(result[0]['id'], 3)
            self.assertEqual(result[1]['id'], 1)
            self.assertEqual(result[2]['id'], 2)

    def test_get_filter_by(self):
        with self.app.test_client() as c:
            todoOne = {'title': 'before', 'description': 'test description1', 'status': 'test'}
            todoTwo = {'title': 'after', 'description': 'test description2', 'status': 'test'}
            todoThree = {'title': 'title', 'description': 'test description3', 'status': 'close'}
            rv = c.post('/api/v1/todo', data=todoOne)
            rv = c.post('/api/v1/todo', data=todoTwo)
            rv = c.post('/api/v1/todo', data=todoThree)
            rv = c.get('/api/v1/todos?filter=id%3D2')
            self.assertEqual(rv.status_code, 200)
            result = json.loads(rv.data)
            self.assertEqual(len(result), 1)
            self.assertEqual(result[0]['id'], 2)
            rv = c.get('/api/v1/todos?filter=id%3D2')
            self.assertEqual(rv.status_code, 200)
            result = json.loads(rv.data)
            self.assertEqual(len(result), 1)
            self.assertEqual(result[0]['id'], 2)
            rv = c.get('/api/v1/todos?filter=status%3D%27test%27%20and%20id%3C2')
            self.assertEqual(rv.status_code, 200)
            result = json.loads(rv.data)
            self.assertEqual(len(result), 1)
            self.assertEqual(result[0]['id'], 1)


if __name__ == '__main__':
    unittest.main()


sample/sample0_test.py
import requests
domain = "http://127.0.0.1:5000/api/v1"

req = requests.get("%s/todos" % domain)
print(req.status_code)
print(req.text)
print(req.headers['content-type'])

req = requests.post("%s/todo" % domain, data={'title': 'supertitle', 'description': 'superdescription'})
print(req.status_code)
print(req.text)
print(req.headers['content-type'])

req = requests.get("%s/todo/1" % domain)
print(req.status_code)
print(req.text)
print(req.headers['content-type'])

req = requests.put("%s/todo/1" % domain, data={'id': 1, 'title': 'supertitle vraiment super', 'description': 'superdescription super', 'status': 'open'})
print(req.status_code)
print(req.text)
print(req.headers['content-type'])

req = requests.patch("%s/todo/1" % domain, data={'title': 'title', 'status': 'closed'})
print(req.status_code)
print(req.text)
print(req.headers['content-type'])

req = requests.get("%s/todo/1" % domain)
print(req.status_code)
print(req.text)
print(req.headers['content-type'])

req = requests.delete("%s/todo/1" % domain)
print(req.status_code)
print(req.text)
print(req.headers['content-type'])


sample/sample1_test.py
import requests
domain = "http://127.0.0.1:5000/api/v1"

req = requests.get("%s/todos" % domain, headers={"Authorization": "Basic tokenadmin"})
print(req.status_code)
print(req.text)
print(req.headers['content-type'])

req = requests.post("%s/todo" % domain, data={'title': 'supertitle', 'description': 'superdescription'}, headers={"Authorization": "Basic tokenadmin"})
print(req.status_code)
print(req.text)
print(req.headers['content-type'])

req = requests.get("%s/todo/1" % domain, headers={"Authorization": "Basic tokenadmin"})
print(req.status_code)
print(req.text)
print(req.headers['content-type'])

req = requests.put("%s/todo/1" % domain, data={'id': 1, 'title': 'supertitle vraiment super', 'description': 'superdescription super', 'status': 'open'}, headers={"Authorization": "Basic tokenadmin"})
print(req.status_code)
print(req.text)
print(req.headers['content-type'])

req = requests.patch("%s/todo/1" % domain, data={'title': 'title', 'status': 'closed'}, headers={"Authorization": "Basic tokenadmin"})
print(req.status_code)
print(req.text)
print(req.headers['content-type'])

req = requests.get("%s/todo/1" % domain, headers={"Authorization": "Basic tokenadmin"})
print(req.status_code)
print(req.text)
print(req.headers['content-type'])

req = requests.delete("%s/todo/1" % domain, headers={"Authorization": "Basic tokenadmin"})
print(req.status_code)
print(req.text)
print(req.headers['content-type'])


sample/sample0.py
import os
import logging
from flask import Flask
from flask_sqlalchemy import SQLAlchemy
from flask_sqlalchemy_api import ApiRest, error_api

app = Flask(__name__)

# db SQLAlchemy
database_file = "sqlite:///{}".format(os.path.join('.', "test.db"))
app.config["SQLALCHEMY_DATABASE_URI"] = database_file
app.config["SQLALCHEMY_TRACK_MODIFICATIONS"] = False
db = SQLAlchemy()


class Todo(db.Model):
    __tablename__ = 'todo'

    __table_args__ = (
        db.Index(
            'ix_unique_index',  # Index name
            'title', 'description',  # Columns which are part of the index
            unique=True),  # The condition
    )

    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
    title = db.Column(db.String, nullable=False)
    description = db.Column(db.String, nullable=True)
    status = db.Column(db.String, nullable=True)


apiManager = ApiRest(db)
for method in ['ALL', 'POST', 'GET', 'DELETE', 'PUT', 'PATCH']:
    apiManager.add_api(Todo, method)
app.register_blueprint(apiManager)


if __name__ == "__main__":
    db.init_app(app)
    with app.app_context():
        db.create_all()
    app.logger.setLevel(logging.DEBUG)
    app.run(host='0.0.0.0', port=5000, debug=True)


flask_sqlalchemy_api/main.py
from flask import Blueprint, request, current_app
from flask_sqlalchemy import SQLAlchemy
from sqlalchemy import PrimaryKeyConstraint, UniqueConstraint, Integer, Float
from sqlalchemy.sql import text
from sqlalchemy.orm.exc import UnmappedInstanceError
from functools import wraps
import logging

__version__ = '0.9.1'

SQLTYPE_TO_FLASKTYPE = {Integer: 'int', Float: 'float'}


def model_to_dict(obj):
    return {c.name: getattr(obj, c.name) for c in obj.__table__.columns}


def unique_endpoint(cls):
    constraint = None
    if len([constraint for constraint in cls.__table__.constraints if isinstance(constraint, UniqueConstraint)]) > 0:
        constraint = [constraint for constraint in cls.__table__.constraints if isinstance(constraint, UniqueConstraint)][0]
    if len([constraint for constraint in cls.__table__.constraints if isinstance(constraint, PrimaryKeyConstraint)]) > 0:
        constraint = [constraint for constraint in cls.__table__.constraints if isinstance(constraint, PrimaryKeyConstraint)][0]
    if constraint is None:
        raise "%s does not PrimaryKeyConstraint or UniqueConstraint" % cls.__name__
    return ("/".join(["<%s:%s>" % (SQLTYPE_TO_FLASKTYPE.get(column.type.__class__, 'string'), column.name) for column in constraint.columns]))


class ItemNotFound(Exception):

    def __init__(self, item):
        super().__init__("Item not found %s" % item)


def error_api(func):
    @wraps(func)
    def decorated_view(*args, **kwargs):
        try:
            return func(*args, **kwargs)
        except UnmappedInstanceError as err:
            return {"code": 400, 'description': 'Item not found, %s' % str(err)}, 400
        except Exception as err:
            return {"code": 400, 'description': str(err)}, 400
    return decorated_view


def multi_decorators(decorators):
    def decorator(f):
        for d in reversed(decorators):
            f = d(f)
        return f
    return decorator


class ApiRest(Blueprint):

    def __init__(self, db, name='apirest', import_name=__name__, url_prefix='/api/v1', *args, **kwargs):
        Blueprint.__init__(self, name, import_name, url_prefix, *args, **kwargs)
        self._db = db
        self._url_prefix = url_prefix

    def add_api(self, cls, method, decorators=[], endpoint=None, serialize=model_to_dict):
        decorators = [error_api,] + decorators
        if method not in ['ALL', 'POST', 'GET', 'DELETE', 'PUT', 'PATCH']:
            raise "%s is not a method value (ALL, POST, GET, DELETE, PUT, PATCH)" % method
        if method == 'ALL':
            method = 'GET'
            if endpoint is None:
                endpoint = '%s/%ss' % (self._url_prefix, cls.__name__.lower())
            self.add_url_rule(endpoint, 'view_%ss' % endpoint[1:], multi_decorators(decorators)(self._all(cls, serialize)), methods=[method, ])
        elif method == 'POST':
            if endpoint is None:
                endpoint = '%s/%s' % (self._url_prefix, cls.__name__.lower())
            self.add_url_rule(endpoint, 'post_%s' % endpoint[1:], multi_decorators(decorators)(self._post(cls, serialize)), methods=[method, ])
        elif method == 'GET':
            if endpoint is None:
                endpoint = '%s/%s/%s' % (self._url_prefix, cls.__name__.lower(), unique_endpoint(cls))
            self.add_url_rule(endpoint, 'get_%s' % endpoint[1:], multi_decorators(decorators)(self._get(cls, serialize)), methods=[method, ])
        elif method == 'DELETE':
            if endpoint is None:
                endpoint = '%s/%s/%s' % (self._url_prefix, cls.__name__.lower(), unique_endpoint(cls))
            self.add_url_rule(endpoint, 'del_%s' % endpoint[1:], multi_decorators(decorators)(self._del(cls, serialize)), methods=[method, ])
        elif method == 'PUT':
            if endpoint is None:
                endpoint = '%s/%s/%s' % (self._url_prefix, cls.__name__.lower(), unique_endpoint(cls))
            self.add_url_rule(endpoint, 'put_%s' % endpoint[1:], multi_decorators(decorators)(self._put(cls, serialize)), methods=[method, ])
        elif method == 'PATCH':
            if endpoint is None:
                endpoint = '%s/%s/%s' % (self._url_prefix, cls.__name__.lower(), unique_endpoint(cls))
            self.add_url_rule(endpoint, 'patch_%s' % endpoint[1:], multi_decorators(decorators)(self._patch(cls, serialize)), methods=[method, ])
        logging.getLogger("werkzeug").info(" * add url rule %s for %s" % (endpoint, method))

    def _all(self, cls, serialize):
        def fct():
            offset = request.args.get('offset', 0)
            limit = request.args.get('limit', 999)
            order_by = request.args.get('orderby', '')
            filter = request.args.get('filter', '')
            return [serialize(item) for item in cls.query.filter(text(filter)).order_by(text(order_by)).offset(offset).limit(limit).all()], 200
        return fct

    def _post(self, cls, serialize):
        def fct(**kws):
            dct = {key: request.form.get(key) for key in request.form}
            for col in [c.name for c in cls.__table__.columns if c.autoincrement is True]:
                if col in dct:
                    del dct[col]
            item = cls(**dct)
            self._db.session.add(item)
            self._db.session.commit()
            return serialize(item), 201
        return fct

    def _get(self, cls, serialize):
        def fct(**kws):
            item = self._db.one_or_404(self._db.select(cls).filter_by(**kws), description=f"{cls.__name__} with parameters {','.join(['%s:%s' % (kw, kws[kw]) for kw in kws])}.")
            return serialize(item), 200
        return fct

    def _del(self, cls, serialize):
        def fct(**kws):
            item = self._db.one_or_404(self._db.select(cls).filter_by(**kws), description=f"{cls.__name__} with parameters {','.join(['%s:%s' % (kw, kws[kw]) for kw in kws])}.")
            self._db.session.delete(item)
            self._db.session.commit()
            return {"code": 200, "message": "element remove with success", "instance": request.path}, 200
        return fct

    def _put(self, cls, serialize):
        def fct(**kws):
            item = self._db.one_or_404(self._db.select(cls).filter_by(**kws), description=f"{cls.__name__} with parameters {','.join(['%s:%s' % (kw, kws[kw]) for kw in kws])}.")
            dct = {key: request.form.get(key) for key in request.form}
            for key in kws:
                dct[key] = kws[key]
            for col in [c.name for c in item.__table__.columns]:
                item.__setattr__(col, dct.get(col))
            self._db.session.commit()
            return serialize(item), 200
        return fct

    def _patch(self, cls, serialize):
        def fct(**kws):
            item = self._db.one_or_404(self._db.select(cls).filter_by(**kws), description=f"{cls.__name__} with parameters {','.join(['%s:%s' % (kw, kws[kw]) for kw in kws])}.")
            dct = {key: request.form.get(key) for key in request.form}
            for key in kws:
                dct[key] = kws[key]
            for col in [key for key in dct if key in item.__table__.columns]:
                item.__setattr__(col, dct.get(col))
            self._db.session.commit()
            return serialize(item), 200
        return fct

    def register(self, app, options):
        try:
            Blueprint.register(self, app, options)
        except Exception:
            app.logger.error("init ApiRest on register is failed")


sample/sample1.py
import os
import logging
from flask import Flask, request, json, make_response
from flask_sqlalchemy import SQLAlchemy
from flask_sqlalchemy_api import ApiRest, error_api
# ------------------------ Manage login ----------------------------
from flask_login import LoginManager, login_user, logout_user, current_user, login_required
from flask import abort
from werkzeug.security import generate_password_hash, check_password_hash
import base64

app = Flask(__name__)

# db SQLAlchemy
database_file = "sqlite:///{}".format(os.path.join('.', "test.db"))
app.config["SQLALCHEMY_DATABASE_URI"] = database_file
app.config["SQLALCHEMY_TRACK_MODIFICATIONS"] = False
db = SQLAlchemy()

# ------------------------ Manage login ----------------------------
login_manager = LoginManager()
login_manager.init_app(app)
login_manager.login_view = 'login'
app.secret_key = 'my_secret_key'


class User(db.Model):
    __tablename__ = 'user'

    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
    name = db.Column(db.String, nullable=False, unique=True)
    password = db.Column(db.String, nullable=False)
    apikey = db.Column(db.String, nullable=True)
    token = db.Column(db.String, nullable=True)

    def __setattr__(self, name, value):
        if name == 'password':
            value = generate_password_hash(value)
        db.Model.__setattr__(self, name, value)

    def __getattribute__(self, name):
        if name not in ('id') and db.Model.__getattribute__(self, name) is None:
            return ""
        return db.Model.__getattribute__(self, name)

    def is_active(self):
        """True, as all users are active."""
        return True

    def get_id(self):
        """Return the id to satisfy Flask-Login's requirements."""
        return self.id

    def is_anonymous(self):
        """False, as anonymous users aren't supported."""
        return False

    def is_authenticated(self):
        return True

    def check_password(self, password):
        return check_password_hash(self.password, password)


@login_manager.unauthorized_handler
def unauthorized():
    return {'authenticated': False}, 401


@login_manager.user_loader
def user_loader(id):
    return User.query.get(id)


@login_manager.request_loader
def load_user_from_request(request):
    # first, try to login using the api_key url arg
    apikey = request.args.get('api')
    if apikey:
        user = User.query.filter_by(apikey=apikey).first()
        if user is not None:
            return user

    # next, try to login using Basic Auth
    token = request.headers.get('Authorization')
    if token:
        token = token.replace('Basic ', '', 1)
        try:
            token = base64.b64decode(token)
        except Exception:
            pass
        user = User.query.filter_by(token=token).first()
        if user is not None:
            return user

    # finally, return None if both methods did not login the user
    return None


@app.route('/login', methods=['POST'])
def login():
    username = request.form['username']
    password = request.form['password']
    registered_user = User(username)
    if registered_user.check_password(password):
        login_user(registered_user, remember=True)
        return {'authenticated': True}, 200
    return {'authenticated': False}, 401


@app.route('/logout')
def logout():
    logout_user()
    return {}, 200

# ------------------------ End Manage login ----------------------------


class Todo(db.Model):
    __tablename__ = 'todo'

    __table_args__ = (
        db.Index(
            'ix_unique_index',  # Index name
            'title', 'description',  # Columns which are part of the index
            unique=True),  # The condition
    )

    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
    title = db.Column(db.String, nullable=False)
    description = db.Column(db.String, nullable=True)
    status = db.Column(db.String, nullable=True)


apiManager = ApiRest(db)
for cls in [Todo, User]:
    for method in ['ALL', 'POST', 'GET', 'DELETE', 'PUT', 'PATCH']:
        apiManager.add_api(cls, method, decorators=[login_required])
app.register_blueprint(apiManager)


if __name__ == "__main__":
    db.init_app(app)
    with app.app_context():
        db.create_all()
        if len(User.query.all()) == 0:
            u = User(name="admin", password="admin", apikey="apikeyadmin", token="tokenadmin")
            db.session.add(u)
            db.session.commit()
    app.logger.setLevel(logging.DEBUG)
    app.run(host='0.0.0.0', port=5000, debug=True)
